File: benchmark/template/programs/swift/micro/mb01.swift
// MB-01 - Fibonacci: naive double recursion, n = 30...37 inclusive.
#if canImport(Darwin)
import Darwin  // fflush(stdout) for the section 5.2 ITER lines
#else
import Glibc   // the same fflush(stdout) on Linux, where the benchmark runs
#endif

func fib(_ n: Int) -> Int {
    if n < 2 {
        return n
    }
    return fib(n - 1) + fib(n - 2)
}

/// The whole workload body (methodology 06 section 4, MB-01).
func workload() -> String {
    var total = 0
    for n in 30...37 {
        total += fib(n)
    }
    return "MB01 \(total)"
}

// Section 5.2 program modes: argv[1] in {once, steady}, argv[2] = K (default 7).
func main() {
    let args = CommandLine.arguments
    let mode = args.count > 1 ? args[1] : "once"
    var line: String
    if mode == "steady" {
        var iterations = args.count > 2 ? (Int(args[2]) ?? 7) : 7
        if iterations < 1 {
            iterations = 7
        }
        let clock = ContinuousClock()
        line = ""
        for k in 0..<iterations {
            let elapsed = clock.measure {
                line = workload()
            }
            let parts = elapsed.components
            let ns = parts.seconds * 1_000_000_000 + parts.attoseconds / 1_000_000_000
            print("ITER \(k) \(ns)")
            fflush(stdout)  // section 5.2: each ITER line is flushed immediately
        }
    } else {
        line = workload()
    }
    print(line)
}

main()

File: benchmark/template/programs/zig/micro/mb01.zig
// MB-01 - Fibonacci: naive double recursion, n = 30..37 inclusive.
const std = @import("std");

fn fib(n: u32) u64 {
    if (n < 2) return n;
    return fib(n - 1) + fib(n - 2);
}

/// The whole workload body (section 5.2: this is what `steady` re-runs).
fn workload() u64 {
    var total: u64 = 0;
    var n: u32 = 30;
    while (n <= 37) : (n += 1) {
        total += fib(n);
    }
    return total;
}

pub fn main(init: std.process.Init.Minimal) !void {
    var threaded = std.Io.Threaded.init_single_threaded;
    const io = threaded.io();
    const stdout = std.Io.File.stdout();

    // Section 5.2 program modes: argv[1] in {once, steady}, argv[2] = K (default 7).
    var mode: []const u8 = "once";
    var k_count: usize = 7;
    var args = init.args.iterate();
    _ = args.next();
    if (args.next()) |a| {
        mode = a;
        if (args.next()) |b| k_count = std.fmt.parseInt(usize, b, 10) catch 7;
    }
    if (k_count < 1) k_count = 7;

    var total: u64 = 0;
    if (std.mem.eql(u8, mode, "steady")) {
        var ibuf: [64]u8 = undefined;
        for (0..k_count) |k| {
            // Section 5.2 monotonic clock (Zig 0.16: std.Io.Clock.awake, see notes below).
            const t0 = std.Io.Clock.awake.now(io);
            total = workload();
            const t1 = std.Io.Clock.awake.now(io);
            std.mem.doNotOptimizeAway(total);
            const ns: u64 = @intCast(t0.durationTo(t1).nanoseconds);
            const iter_line = try std.fmt.bufPrint(&ibuf, "ITER {d} {d}\n", .{ k, ns });
            // stdout is written unbuffered, so every ITER line is flushed as it is produced.
            try stdout.writeStreamingAll(io, iter_line);
        }
    } else {
        total = workload();
    }

    var buf: [64]u8 = undefined;
    const line = try std.fmt.bufPrint(&buf, "MB01 {d}\n", .{total});
    try stdout.writeStreamingAll(io, line);
}
